zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineCache.kt

```diff
@@ -121,28 +117,40 @@ class ZiplineCache internal constructor(
     applicationName: String,
     sha256: ByteString,
     nowEpochMs: Long,
-    download: suspend () -> ByteString?,
-  ): ByteString? {
-    val read = read(sha256, nowEpochMs)
-    if (read != null) return read
+    download: suspend () -> ByteString,
+  ): ByteString {
+    if (hasWriteFailures) return download()
+
+    try {
+      val read = read(sha256, nowEpochMs)
+      if (read != null) return read
+    } catch (e: Exception) {
+      hasWriteFailures = true // Mark this cache as broken.
+      return download()
+    }
+
+    val content = download()
+    try {
+      write(
+        applicationName = applicationName,
+        sha256 = sha256,
+        content = content,
+        isManifest = false,
+        nowEpochMs = nowEpochMs,
+      )
+    } catch (e: Exception) {
```

Comment:
Discarded exception is recovered here:
https://github.com/cashapp/zipline/pull/1375